internal/controller/pathwaysjob_controller.go

```diff
@@ -0,0 +1,663 @@
+
+	if err != nil {
+		return err
+	} else {
+		log2.Info("PathwaysJob: successfully created JobSet: ", "JobSet name", js.Name)
+	}
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PathwaysJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	if err := jobsetv1alpha2.AddToScheme(mgr.GetScheme()); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&pathwaysjob.PathwaysJob{}).
+		Owns(&jobsetv1alpha2.JobSet{}). // PathwaysJob owns the underlying JobSet object
+		Complete(r)
+}
+
+// Find the status of the underlying JobSet for 'colocated' or 'default' deployment modes.
+func (r *PathwaysJobReconciler) findJobSetStatus(ctx context.Context, js *jobsetv1alpha2.JobSet) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("PathwaysJob findJobSetStatus", "Jobset name", js.ObjectMeta.Name)
+
+	for _, c := range js.Status.Conditions {
+		log.Info("\n\n PathwaysJob CONDITION", "CONDITION", c.Type, "Status", c.Status, "Message", c.Message)
+		if (c.Type == string(jobsetv1alpha2.JobSetSuspended) || c.Type == string(jobsetv1alpha2.JobSetCompleted) || c.Type == string(jobsetv1alpha2.JobSetFailed)) && c.Status == metav1.ConditionTrue {
+			log.Info("\n\n PathwaysJob: JobSet in TERMINAL STATE", "Condition ", c.Type)
+		}
+		if (c.Type == string(jobsetv1alpha2.JobSetStartupPolicyCompleted) ||
+			c.Type == string(jobsetv1alpha2.JobSetStartupPolicyInProgress)) && c.Status == metav1.ConditionTrue {
+			log.Info("\n\n PathwaysJob: JobSet in TERMINAL STATE", "Condition ", c.Type)
+		}
+	}
+}
+
+// ---------------------- PATHWAYS HELPERS --------------------------
+// Find TPU version from the worker's type (- used to determine Pathways instance_type)
+func constructTPUVersionFromWorkerType(tpuGKEAcceleratorType pathwaysjob.WorkerType) string {
+	// Worker types are already validated in the YAML.
+	return WorkerTypeToTPUVersionMap[string(tpuGKEAcceleratorType)]
+}
+
+// Validate that topology provided is valid for the provided worker type.
+func validateTPUTopologyWithWorkerType(ctx context.Context, tpuGKEAcceleratorType pathwaysjob.WorkerType, topology string) (string, error) {
+	log := ctrl.LoggerFrom(ctx)
+	if slices.Contains(ValidTpuTopologiesMap[string(tpuGKEAcceleratorType)], topology) {
+		return topology, nil
+	} else {
+		log.Info("Invalid topology!!! ", "Worker type ", string(tpuGKEAcceleratorType), " cannot have topology ", topology)
+		return "", fmt.Errorf("invalid TPU topology for worker type")
+	}
+}
+
+// Calculate the number of VMs based on the Topology (- used in completions/parallelisms)
+func calculateVMsFromTopology(topology string) int32 {
+	parts := strings.Split(topology, "x") // Examples - 2x2x4 or 4x4
+	if len(parts) < 2 {
+		return 0
+	}
+	// Calculate the number of chips based on the Topology.
+	chips := 1
+	for _, part := range parts {
+		num, _ := strconv.Atoi(part)
+		chips *= num
+	}
+	vms := 1
+	chipsperVM := 4
```

Comment:
Just wanted to check this does not need to be architecture specific?

Reply:
At the moment, I added support for machines with -4t architecture, keeping things inline with XPK. I added a ToDo to extend it to machines with -8t architecture. Thanks for the comment!

Reply:
Thank you! That sounds good to me.